cachito/web/content_manifest.py

```diff
@@ -232,3 +238,208 @@ def generate_icm(self, image_contents=None):
         icm = deepcopy(BASE_ICM)
         icm["image_contents"] = image_contents or []
         return deep_sort_icm(icm)
+
+
+@dataclass(frozen=True)
+class Package:
+    """
+    A package within a content manifest.
+
+    It is used primarily to generate a package URL (purl).
+    """
+
+    name: str
+    type: str
+    version: str
+    dev: bool = False
+    dependencies: list = field(default_factory=list)
+    path: str = None
+
+    def __repr__(self):
+        return (
+            f"<{self.__class__.__name__} name={self.name}, type={self.type}, "
+            f"version={self.version}>"
+        )
+
+    def __hash__(self):
+        return hash((self.name, self.type, self.version, self.dev))
+
+    def __eq__(self, other):
+        return (
+            type(other) == Package
+            and self.name == other.name
+            and self.type == other.type
+            and self.version == other.version
+            and self.dev == other.dev
+        )
+
+    @classmethod
+    def from_json(cls, package):
+        """
+        Create a Package object from JSON.
+
+        All dependencies will also be converted to Package objects.
+
+        :param dict package: the dictionary representing the package
+        :return: the Package object
+        :rtype: Package
+        """
+        dependencies = [
+            Package.from_json(dependency) for dependency in package.get("dependencies", [])
+        ]
+
+        return cls(
+            name=package.get("name"),
+            type=package.get("type"),
+            version=package.get("version"),
+            dev=package.get("dev", False),
+            dependencies=dependencies,
+            path=package.get("path"),
```

Comment:
In `packages.json`, the `path` is not present if it is `os.curdir`. For a loaded object from the file, I think we can assign `os.curdir` to the `path` rather than `None`.

Reply:
I don't know if I got it right, but I tried assigning instead of none and the purls started to be generated with a trailing `.`, like this:

`pkg:golang/example.com%2Forg%2Fproject@1.1.1#.`

(which broke a bunch of tests :laughing: )

Reply:
The code expects that packages without a subpath will have `path = None` rather than `path = "."`, so  using None was correct